Add null_safe_string_view(const char*, std::size_t)

alt-graph opened the pull request:
There are use cases for constructing a `string_view` with a specified size in a way that is protected against null pointers. This is notably different from the existing `safe_string_view(const char*, std::size_t)` which terminates the string at a null byte.

alt-graph committed:
Add null_safe_string_view(const char*, std::size_t)

[why]
There are use cases for constructing a string_view with a specified size
that is still protected against null pointers. This is different from
safe_string_view(const char*, std::size_t) which terminates the string
at a null byte.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Update changelog with latest null_safe_string_view() changes

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

Finii commented:
Ugg, this 'Pull Request Overview' really puts me off. I prefer a human generated explanation/overview, and this is just babble. If you absolutely need an LLM "bug finder", maybe we/you can at least re remove the blah as that increases the mental load any PR anyhow has.

To the topic...

a) A wonder if we can not somehow unify `save_string_view()` with the new functionality presented here with `null_safe_string_view()`. I will add something in the bottom.

For me, having 4 functions for "the same" is too much. Even 3 is too much and I must have overlooked it.

b) I imagine that sometimes you want to do something like this (feature idea):

```c++
    auto& setting_a = obj.settings.a;
    std::cout << null_save_string_view(setting_a.charptr, setting_a.length, "Option A unset");
```

---

The current documentation is already wrong:

<img width="1167" height="695" alt="Screenshot 2025-12-15 at 19 39 08" src="https://github.com/user-attachments/assets/5cc52ef1-c55c-4401-a11f-78e2de4c07c2" />

_Doxygen docu has copy&paste error?_

Hmm, I wanted to compare the existing `*safe_string*()` functions. But that seems hard. It is too much cuddle muddle for me.

> This is notably different from the existing safe_string_view(const char*, std::size_t) which terminates the string at a null byte.

And that null-byte give the `null_` in `null_save_string_view()`? I find that not really plausible/catchy. Especially as it also 'could' mean nullptr 😬 .

Hmm, still trying to come up with a feature table

| | creates string_view | can handle `nullptr` | preserves '/0' |
|---|---|---|---|
| safe_string() | ➖  | 👍 | 👎 |
| safe_string_view() | ➕ | 👍 | 👎 |
| null_safe_string() | ➖  | 👍 | 👍 |
| _null_safe_string_view()_ | ➕  | 👍 | 👍 |

So this looks like a reasonable extension/completion. But I believe we should have only 2 functions and not 4, and that the 'preserves `'/0'` should be handled differently (e.g. as nttp or optional parameter).

Hmm. The existing 3 functions are ... functions. Being in this library it means they can not be optimized away, I assume? Maybe they should be templates instead. That is what I believed they were.

Hmm:

* https://github.com/gul-cpp/gul17/commit/050d97ba41be971807c2100737e133f540688e33 `Rename the new functions to null_safe_*()`

Uggg:

* https://github.com/gul-cpp/gul17/pull/29

It seems the whole feature and naming is more complicated than one would expect.
I think we should discuss this a bit more rigorously.
Obviously my table is not sufficient. A good tabe would also help users choose the right of the multiple functions in this family.

Maybe I'm wrong with this, but for me this feels like a thicket.

_Edit: I guess by b) is too much (at least right now)_

alt-graph commented:
> Ugg, this 'Pull Request Overview' really puts me off. I prefer a human generated explanation/overview, and this is just babble. If you absolutely need an LLM "bug finder", maybe we/you can at least re remove the blah as that increases the mental load any PR anyhow has.
> 
> To the topic...
> 
> a) A wonder if we can not somehow unify `save_string_view()` with the new functionality presented here with `null_safe_string_view()`. I will add something in the bottom.
> 
> For me, having 4 functions for "the same" is too much. Even 3 is too much and I must have overlooked it.
> 
> b) I imagine that sometimes you want to do something like this (feature idea):
> 
> ```c++
>     auto& setting_a = obj.settings.a;
>     std::cout << null_save_string_view(setting_a.charptr, setting_a.length, "Option A unset");
> ```
> 
> The current documentation is already wrong:
> <img alt="Screenshot 2025-12-15 at 19 39 08" width="1167" height="695" src="https://private-user-images.githubusercontent.com/16012374/526715229-5cc52ef1-c55c-4401-a11f-78e2de4c07c2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU4NzMwMjYsIm5iZiI6MTc2NTg3MjcyNiwicGF0aCI6Ii8xNjAxMjM3NC81MjY3MTUyMjktNWNjNTJlZjEtYzU1Yy00NDAxLWExMWYtNzhlMmRlNGMwN2MyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE2VDA4MTIwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3ZDdlMDU1Y2M1NmE5NjM3ZDBmYTkzOGZhZjRhMmZkZmQ4NGVmZGYwNDU0M2Q0YjFmYWIyMjIwMjA5NmQxMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Mx1uWYofPfeDOruvUXaeXyl_fV8ocv-H8vq2N6vFvRs">
> 
> _Doxygen docu has copy&paste error?_
> 
> Hmm, I wanted to compare the existing `*safe_string*()` functions. But that seems hard. It is too much cuddle muddle for me.
> 
> > This is notably different from the existing safe_string_view(const char*, std::size_t) which terminates the string at a null byte.
> 
> And that null-byte give the `null_` in `null_save_string_view()`? I find that not really plausible/catchy. Especially as it also 'could' mean nullptr 😬 .
> 
> Hmm, still trying to come up with a feature table
> 	creates string_view 	can handle `nullptr` 	preserves '/0'
> safe_string() 	➖ 	👍 	👎
> safe_string_view() 	➕ 	👍 	👎
> null_safe_string() 	➖ 	👍 	👍
> _null_safe_string_view()_ 	➕ 	👍 	👍
> 
> So this looks like a reasonable extension/completion. But I believe we should have only 2 functions and not 4, and that the 'preserves `'/0'` should be handled differently (e.g. as nttp or optional parameter).
> 
> Hmm. The existing 3 functions are ... functions. Being in this library it means they can not be optimized away, I assume? Maybe they should be templates instead. That is what I believed they were.
> 
> Hmm:
> 
>     * [050d97b](https://github.com/gul-cpp/gul17/commit/050d97ba41be971807c2100737e133f540688e33) `Rename the new functions to null_safe_*()`
> 
> 
> Uggg:
> 
>     * [Add single-argument overloads safe_string(const char*) and safe_string_view(const char*) #29](https://github.com/gul-cpp/gul17/pull/29)
> 
> 
> It seems the whole feature and naming is more complicated than one would expect. I think we should discuss this a bit more rigorously. Obviously my table is not sufficient. A good tabe would also help users choose the right of the multiple functions in this family.
> 
> Maybe I'm wrong with this, but for me this feels like a thicket.
> 
> _Edit: I guess by b) is too much (at least right now)_

I know that naming is complicated, but... let me just point out that the functions that are now called `null_safe_string*()` were originally proposed under the name `safe_string*()`, and a certain @Finii [argued](https://github.com/gul-cpp/gul17/pull/29) that they should have a different name. :)

Also, why is "the current documentation [...] already wrong"? I do not see that. The "null_safe_" functions only protect against null pointers, the "safe_" functions also terminate the string early at a null byte. In that regard, the new `safe_string_view()` overload is just the one missing extension.

We can discuss if we should inline these functions, but IMO that should be a different MR because we would also touch the existing functions.

Finii commented:
> I know that naming is complicated, but... let me just point out that the functions that are now called `null_safe_string*()` were originally proposed under the name `safe_string*()`, and a certain @Finii [argued](https://github.com/gul-cpp/gul17/pull/29) that they should have a different name. :)

I also mentioned that?

> Also, why is "the current documentation [...] already wrong"? I do not see that.

The examples?

<img width="1277" height="1018" alt="image" src="https://github.com/user-attachments/assets/424c6b82-7154-4dc7-b51c-7c763c0acb36" />


> The "null_safe_" functions only protect against null pointers, the "safe_" functions also terminate the string early at a null byte. In that regard, the new `safe_string_view()` overload is just the one missing extension.

I just suggested putting this kind of explanation into the docs. What `save_` and `null_` means, such that one can more easily 'assemble' the function name to search for ;)

----

I guess the change is ok, but I would like to see the documentation fixed (Soeren would fix them in an additional PR, but I would fix them en passant). And maybe an expansion in the documentation similar to what you gave above.

alt-graph committed:
Improve documentation of safe_string*(), null_safe_string*() functions

[why]
It is a jungle. The safe_string*() functions respect C-string style null
termination, the null_safe_string*() functions only do that if they do
not have a length parameter.

Proposed-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
I have fixed the Doxygen examples for two of the functions and tried to make the "first-line" description of each function as explicit as possible. Have another look – it really is an unpleasant thicket of functions now.

alt-graph commented:
Force-pushed with a whitespace change.

Finii commented:
> Have another look – it really is an unpleasant thicket of functions now.

Maybe we should try a more organized approach and clean this up. Next year. If time permits :rofl:
